plugins/inputs/slurm/slurm.go

```diff
@@ -0,0 +1,508 @@
+	"context"
+	stdTls "crypto/tls"
+	_ "embed"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+
+	goslurm "github.com/pcolladosoto/goslurm/v0038"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type Slurm struct {
+	URL              string          `toml:"url"`
+	Username         string          `toml:"username"`
+	Token            string          `toml:"token"`
+	IgnoredEndpoints []string        `toml:"ignored_endpoints"`
+	ResponseTimeout  config.Duration `toml:"response_timeout"`
+	tls.ClientConfig
+
+	client      *goslurm.APIClient
+	baseURL     *url.URL
+	endpointMap map[string]bool
+}
+
+func (*Slurm) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *Slurm) Init() error {
+	if s.ResponseTimeout == 0 {
+		s.ResponseTimeout = config.Duration(time.Second * 5)
+	}
+
+	s.endpointMap = map[string]bool{}
+	for _, endpoint := range s.IgnoredEndpoints {
+		s.endpointMap[strings.ToLower(endpoint)] = true
+	}
+
+	if s.URL == "" {
+		return errors.New("empty URL provided")
+	}
+
+	u, err := url.Parse(s.URL)
+	if err != nil {
+		return err
+	}
+
+	if u.Hostname() == "" {
+		return fmt.Errorf("empty hostname for url %q", s.URL)
+	}
+
+	s.baseURL = u
+
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return fmt.Errorf("invalid scheme %q", u.Scheme)
+	}
+
+	var tlsCfg *stdTls.Config
+	if u.Scheme == "https" {
+		tlsCfg, err = s.ClientConfig.TLSConfig()
+		if err != nil {
+			return err
+		}
+	}
+
+	configuration := goslurm.NewConfiguration()
+	configuration.Host = u.Host
+	configuration.Scheme = u.Scheme
+	configuration.UserAgent = internal.ProductToken()
+	configuration.HTTPClient = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+		},
+		Timeout: time.Duration(s.ResponseTimeout),
+	}
+
+	s.client = goslurm.NewAPIClient(configuration)
+
+	return nil
+}
+
+func (s *Slurm) parseTres(tres string) map[string]interface{} {
+	parsedValues := map[string]interface{}{}
+	var ok bool
+
+	for _, tresVal := range strings.Split(tres, ",") {
+		parsedTresVal := strings.Split(tresVal, "=")
+		if len(parsedTresVal) != 2 {
+			continue
+		}
+
+		tag := parsedTresVal[0]
+		val := parsedTresVal[1]
+		var factor float64 = 1
+
+		if tag == "mem" {
+			factor, ok = map[string]float64{
+				"K": 1.0 / 1024.0,
+				"M": 1,
+				"G": 1024,
+				"T": 1024 * 1024,
+			}[strings.ToUpper(val[len(val)-1:])]
+			if !ok {
+				continue
+			}
+			val = val[:len(val)-1]
+		}
+
+		parsedInt, err := strconv.ParseInt(val, 10, 64)
+		if err == nil && factor >= 1 {
+			parsedValues[tag] = parsedInt * int64(factor)
+			continue
+		}
+		parsedFloat, err := strconv.ParseFloat(val, 64)
+		if err == nil {
+			parsedValues[tag] = parsedFloat * factor
+			continue
+		}
+		parsedValues[tag] = val
```

Comment:
Are we really sure float values are **never** parsed as here? We had troubles with this kind of "auto-typing" in the past and I don't want to run into this once more...

Reply:
I'm not entirely sure about what you mean... sorry about that! The idea behind this parsing logic is that we'll try to parse the value as an integer and, if that fails we'll try to check whether it's a float. It not, then we'll just use the string 'as-is' without any additional parsing other than converting the comma separated list into a map of keys and values.

Maybe you are referring to whether we might parse something that's not a float as a float if the format lines up?

Anyway, I'd be more than happy to simply work with a `map[string]string` at the expense of losing the type information.

Sorry for not grasping the gist of your comment again...

Reply:
Let me give you an example. Assume we got the following string values for the same `tag` (e.g. `temperature`):
1. `temperature=23.9`
2. `temperature=23.1`
3. `temperature=23`

The first two ones will clearly produce an error when trying to parse them as an `int` so you will get a field named `temperature` with a float value for the first two examples. However, the third line perfectly parses as an integer and therefore you will get a resulting `temperature` field being an**integer** instead of a float. This will lead to type conflicts in e.g. InfluxDB as the same column cannot hold different datatypes.

Thinking more about it, you will have the same issue with
1. `mem=10M` resulting in integer `10`
2. `mem=1G` resulting in integer `1024`
3. `mem=100k` resulting in **float** `0.09765625`

Reply:
Yeah... this type of scenario is why I didn't jump head first into this parsing during the first iteration...

Now, I guess the best thing we can do is to just parse every number as a `float64` and, if anything goes wrong, default to returning a `string` altogether.

Another approach would be to simply work with `string`s altogether and just content ourselves with splitting up the TRES fields on `,` and `=`.

I'm more than happy to provide iterations for both scenarios above, but I'd very much rather hear your thoughts (and insight) beforehand...

Thanks in advance!

Reply:
Hi @srebhan! I changed the parsing logic so that only `float64`s will be parsed if a numerical quantity is found. Otherwise, the raw string will be included as a metric. Would that be okay with you?

We can also just parse everything as strings and call it a day so that we're sure no type conflicts will arise...